compaction/task_manager_module.cc

```diff
@@ -448,7 +449,14 @@ future<> table_cleanup_keyspace_compaction_task_impl::run() {
     co_await wait_for_your_turn(_cv, _current_task, _status.id);
     auto owned_ranges_ptr = compaction::make_owned_ranges_ptr(_db.get_keyspace_local_ranges(_status.keyspace));
     co_await run_on_table("force_keyspace_cleanup", _db, _status.keyspace, _ti, [&] (replica::table& t) {
-        return t.perform_cleanup_compaction(owned_ranges_ptr, tasks::task_info{_status.id, _status.shard});
+        try {
+            return t.perform_cleanup_compaction(owned_ranges_ptr, tasks::task_info{_status.id, _status.shard});
+        } catch (const gate_closed_exception& e) {
```

Comment:
I'm a bit worried that we can swallow an unrelated gate closed exception here.
@Deexie please provide a second opinion.

Reply:
I think that can happen e.g. on shutdown, when we create a cg task (which has its own gate), but the table is dropped. Anyway, there's a more bothering issue:

If in a catch-clause here `t.async_gate().is_closed()`  is true then table `t` was already stopped.
If `t` was stopped it means that `database::tables_metadata::_column_families` does not keep a pointer to `t` anymore, see:
https://github.com/scylladb/scylladb/blob/5995a8ad4a6c3ac4d001dd79ff80d322cb8cfb0e/replica/database.cc#L1140-L1164
where we first call `db.detach_column_family(*table_shards);` and then call `table_shards->stop()`.

Now let's see the implementation of run_on_table:
https://github.com/scylladb/scylladb/blob/5995a8ad4a6c3ac4d001dd79ff80d322cb8cfb0e/compaction/task_manager_module.cc#L193-L207
In line 197 we call `find_column_family` which returns a reference to table kept in `database::tables_metadata::_column_families`. So if `func` preempts and the table is dropped (and stopped) in the middle of it we get dangling reference.

I do not see the mistakes in this reasoning, I also do not see anything that would prevent such an execution. But it would be strange if we've never hit it...

Reply:
I think that what we should do is enter the table's `_async_gate` as soon as we get the reference to it in `run_on_table` (prob need to move the function to class table for that, which seems appropriate).

And then in `table::stop`, we shouldn't `co_await _async_gate.close()`, but rather keep the `future<>` it returns and wait for it only later on, after `co_await parallel_foreach_compaction_group(std::mem_fn(&compaction_group::stop));` which calls `co_await _t._compaction_manager.remove(as_table_state())`.

We Should wait on the compaction_group's `_async_gate .close()` in a similar way, after removing it from the compaction_manager and stopping all ongoing compactions, since `compaction_manager::start_compaction` holds that gate.  Otherwise, we can't stop ongoing compaction while they hold those gates.

Reply:
Proposal:
```diff
diff --git a/compaction/task_manager_module.cc b/compaction/task_manager_module.cc
index 6d7e60241b..f5c1c9ee38 100644
--- a/compaction/task_manager_module.cc
+++ b/compaction/task_manager_module.cc
@@ -193,7 +193,8 @@ future<> run_on_table(sstring op, replica::database& db, std::string keyspace, t
     std::exception_ptr ex;
     tasks::tmlogger.debug("Starting {} on {}.{}", op, keyspace, ti.name);
     try {
-        co_await func(db.find_column_family(ti.id));
+        auto tptr = replica::table_ptr(db, ti.id);
+        co_await func(*tptr);
     } catch (const replica::no_such_column_family& e) {
         tasks::tmlogger.warn("Skipping {} of {}.{}: {}", op, keyspace, ti.name, e.what());
     } catch (...) {
diff --git a/replica/database.hh b/replica/database.hh
index bff06cbd54..cf9b37a9b5 100644
--- a/replica/database.hh
+++ b/replica/database.hh
@@ -1180,6 +1180,33 @@ class table : public enable_lw_shared_from_this<table>
     friend class compaction_group;
 };
 
+// Smart table pointer that keeps the table object alive
+// while it's being accessed asynchronously
+class table_ptr {
+    gate::holder _holder;
+    lw_shared_ptr<table> _table_ptr;
+public:
+    explicit table_ptr(table&);
+    table_ptr(database& db, std::string_view ks, std::string_view name);
+    table_ptr(database& db, const table_id& id);
+
+    const table& operator*() const noexcept {
+        return *_table_ptr;
+    }
+
+    table& operator*() noexcept {
+        return *_table_ptr;
+    }
+
+    const table* operator->() const noexcept {
+        return _table_ptr.operator->();
+    }
+
+    table* operator->() noexcept {
+        return _table_ptr.operator->();
+    }
+};
+
 using user_types_metadata = data_dictionary::user_types_metadata;
 
 using keyspace_metadata = data_dictionary::keyspace_metadata;
diff --git a/replica/table.cc b/replica/table.cc
index 2488a507b7..a29717ef5e 100644
--- a/replica/table.cc
+++ b/replica/table.cc
@@ -67,6 +67,19 @@ static seastar::metrics::label keyspace_label("ks");
 
 using namespace std::chrono_literals;
 
+table_ptr::table_ptr(table& t)
+    : _holder(t.async_gate())
+    , _table_ptr(t.shared_from_this())
+{ }
+
+table_ptr::table_ptr(database& db, std::string_view ks, std::string_view name)
+    : table_ptr(db.find_column_family(ks, name))
+{ }
+
+table_ptr::table_ptr(database& db, const table_id& id)
+    : table_ptr(db.find_column_family(id))
+{ }
+
 void table::update_sstables_known_generation(sstables::generation_type generation) {
     auto gen = generation ? generation.as_int() : 0;
     if (_sstable_generation_generator) {
@@ -1073,10 +1086,14 @@ table::stop() {
     if (_async_gate.is_closed()) {
         co_return;
     }
-    co_await _async_gate.close();
+    // Allow `compaction_manager::stop` to stop ongoing compactions
+    // while they may still hold the _async_gate
+    // entered in `compaction_manager::start_compaction`
+    auto gate_closed_fut = _async_gate.close();
     co_await await_pending_ops();
     co_await parallel_foreach_compaction_group(std::mem_fn(&compaction_group::stop));
     co_await _sstable_deletion_gate.close();
+    co_await std::move(gate_closed_fut);
     co_await get_row_cache().invalidate(row_cache::external_updater([this] {
         for (const compaction_group_ptr& cg : compaction_groups()) {
             cg->clear_sstables();
@@ -1654,9 +1671,13 @@ future<> compaction_group::stop() noexcept {
     if (_async_gate.is_closed()) {
         co_return;
     }
-    co_await _async_gate.close();
+    // Allow `compaction_manager::remove` to stop ongoing compactions
+    // while they may still hold the _async_gate
+    // entered in `compaction_manager::start_compaction`
+    auto gate_closed_fut = _async_gate.close();
     co_await flush();
     co_await _t._compaction_manager.remove(as_table_state());
+    co_await std::move(gate_closed_fut);
 }
 
 void compaction_group::clear_sstables() {
```

Reply:
https://github.com/scylladb/scylladb/pull/16351